design_notes/v2.md

```diff
@@ -0,0 +1,127 @@
+- [Architecture diagram](https://github.com/zarr-developers/VirtualiZarr/issues/225)
+
+## Data structures
+
+- `ChunkManifest`: An object defining the paths, chunk byte offsets, lengths of one virtual array.
+- `ManifestArray`: `ManifestArrays` contain a single chunk manifest and its associated Zarr V3 `ArrayMetadata` and provide a way to concatenate and stack multiple arrays.
+- `ManifestGroup`: `ManifestGroup`s are a way to organize child groups and arrays into a data hierarchy. Contains Zarr V3 `GroupMetadata` and one or more `ManifestGroup`s and/or `ManifestArray`s.
+- `ManifestStore`: An implementation of Zarr's store API that contains a `ManifestGroup` and an `AsyncObjectReader` (see protocol below). The `ManifestStore` redirects incoming `get` requests to the byte ranges contained within the `ChunkManifest`.
+- `xarray.Dataset`: Contains `ManifestArray`-backed variables and (optionally) in-memory NumPy-backed variables.
+
+## Definitions
+
+- `Parser` - A parser is a function that constructs a `ManifestStore` from an `AsyncObjectReader` and a `file_url`.
+
+## Protocols
+
+**Note**: We anticipate using [obspec](https://developmentseed.org/obspec/latest/) with the following typing in the future, but currently use only `ObjectStore` rather than the general interface definition.
+
+```python
+from obspec import GetAsync, GetRangeAsync, List
+
+class AsyncObjectReader(GetAsync, GetRangeAsync, List):
+    ...
+
+```
+
+```python
+class Parser(protocol):
+    """
+    Callback protocol for creating a virtual zarr store from a file
+    """
+    def __call__(
+        self,
+        file_url: str,
+        object_reader: ObjectStore,
+        *,
+    ) → ManifestStore:
+        ...
+```
+
+## Top-level API functions
+
+- `open_virtual_dataset` - construct a Xarray Dataset with NumPy and/or ManifestArray-backed variables from a file
+- `open_virtual_mfdataset` - construct a Xarray Dataset with NumPy and/or ManifestArray-backed variables from many files
+
+> **NOTE**: Here the construction of a virtual dataset from a manifest store would happen in the top-level function rather than the backends.
+
+```python
+ def open_virtual_dataset(
+    file_url: str,
+    object_reader: ObjectStore,
+    parser: Parser,
+    *,
+    drop_variables: Iterable[str] | None = None,
+    loadable_variables: Iterable[str] | None = None,
+    decode_times: bool | None = None,
+    cftime_variables: Iterable[str] | None = None,
+    indexes: Mapping[str, Index] | None = None,
+) → Dataset
+    """
+    Method for creating virtual Xarray dataset, only implemented in top level API
+    """
+    manifest_store = parser(
+        file_url = file_url,
+        object_reader = object_reader,
+    )
+    ds = manifest_store.to_virtual_dataset(
+        loadable_variables=loadable_variables,
+        decode_times=decode_times,
+        indexes=indexes,
+    )
+    return ds
```

Comment:
```suggestion
    vds = manifest_store.to_virtual_dataset(
        loadable_variables=loadable_variables,
        decode_times=decode_times,
        indexes=indexes,
        drop_variables=drop_variables
    )
    return vds
```
@sharkinsspatial @TomNicholas each of the parsers/backends right now has redundant code to deal with `drop_variables`. I think we should instead have `drop_variables` handled either in the top level `open_virtual_dataset` via `return ds.drop_vars(drop_variables)` or in `to_virtual_dataset` which could just skip any variables in `drop_variables`.